tsl/src/remote/dist_copy.c

```diff
@@ -840,34 +1466,59 @@ remote_distributed_copy(const CopyStmt *stmt, CopyChunkState *ccstate, List *att
 												   ht,
 												   GetPerTupleExprContext(estate),
 												   attnums,
-												   copy_should_send_binary());
+												   copy_should_send_binary(stmt));
 	uint64 processed = 0;
 
+	MemoryContext batch_context =
+		AllocSetContextCreate(CurrentMemoryContext, "Remote COPY batch", ALLOCSET_DEFAULT_SIZES);
+
 	PG_TRY();
 	{
+		MemoryContextSwitchTo(batch_context);
 		while (true)
 		{
-			ResetPerTupleExprContext(ccstate->estate);
-			MemoryContextSwitchTo(GetPerTupleMemoryContext(ccstate->estate));
-
 			CHECK_FOR_INTERRUPTS();
 
-			if (!read_next_copy_row(context, ccstate->cstate))
-				break;
+			ResetPerTupleExprContext(ccstate->estate);
 
+			/* Actually process the next row. */
+			bool eof = !read_next_copy_row(context, ccstate->cstate);
+			if (!eof && context->batch_row_count < MAX_BATCH_ROWS &&
+				context->batch_size_bytes < MAX_BATCH_BYTES)
+			{
+				/*
+				 * Accumulate more rows into the current batch.
+				 */
+				continue;
+			}
+
+			/*
+			 * Send out the current batch.
+			 */
 			remote_copy_process_and_send_data(context);
-			++processed;
+
+			processed += context->batch_row_count;
+			context->batch_row_count = 0;
+			context->batch_size_bytes = 0;
+			context->batch_ordinal++;
+			MemoryContextReset(batch_context);
+
+			if (eof)
+			{
+				break;
+			}
 		}
 	}
 	PG_CATCH();
 	{
 		/* If we hit an error, make sure we end our in-progress COPYs */
-		remote_copy_end(context);
+		end_copy_on_failure(&context->connection_state);
```

Comment:
suggest naming this similar/analagous to the success function. Maybe also taking just `context`?

```suggestion
		remote_copy_end_on_failure(context);
```